### PR TITLE
Add support for pvvx mithermometer display via ble client

### PR DESCRIPTION
**Pull request by puuu**
# What does this implement/fix?

Add a pvvx_mithermometer display platform that allows to use devices running the [ATC_MiThermometer firmware](https://github.com/pvvx/ATC_MiThermometer) by pvvx as display drivers with ESPHome.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1986

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml
esp32_ble_tracker:

ble_client:
- mac_address: "A4:C1:38:B1:CD:7F"
  id: pvvx_ble_display

display:
- platform: pvvx_mithermometer
  ble_client_id: pvvx_ble_display
  lambda: |-
    it.print_bignum(23.1);
    it.print_unit(pvvx_mithermometer::UNIT_DEG_C);
    it.print_smallnum(33);
    it.print_percent(true);
    it.print_happy(true);
    it.print_bracket(true);
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by nagyrobi** [comment]
How do you ensure that the screen of the device is not overwirtten with local data?
AFAIK by default `pvvx` firmware Xiaomi Mijia LYWSD03MMC refreshes periodically the screen with its own sensors data. Is there a way for the user to know that the currently displayed information comes from the local sensor of the pvvx device or from a remote sensor through ESPHome?

**Comment by puuu** [comment]
@nagyrobi Yes the pvvx firmware refresh the screen periodically (can be configured as minimum LCD refresh rate). You can configure different modes and the display will change between this modes with the LCD refresh rate, e.g. displaying the battery percentage or a clock. In addition, the pvvx firmware supports external data that is provided via BLE.

You can change all the configurations with a browser: [TelinkMiFlasher.html](https://pvvx.github.io/ATC_MiThermometer/TelinkMiFlasher.html). This tool also allows to set external data. The external data has a  validity time of how long this external data should be shown. The firmware switches then between the different views including the external data.

You can not directly recognizes if the data shown is local or external data, but by changing the smiley (e.g. no brackets) or the units it can be made somehow obviously. Also enabling the clock makes it more  obviously, then the cycles are: local data, clock, external data, clock and so on.

This platform only send the external data (and optional a timestamp) via BLE to the pvvx device. Configuration (like enabling the clock and refresh rate) of the pvvx device must be done by the user via e.g. browser.

I hope this answers your question. Thank you for asking, I will add some more information about this to the documentation.

**Comment by nagyrobi** [comment]
Yes, that's why I was asking. The pvvx repo's documentation is not quite obvious (the guy is not a native English speaker) so imho we should add more detail here, related to this functionality.

**Review by pasiz — APPROVED** [review]
By me looks good so far. Please test it with different configurations to ensure all previous functionality is not affected in any means.

-Pasi

**Comment by puuu** [comment]
@pasiz Thank you. I have this code running for weeks with 4 LYWSD03MMC. The configuration is similar to the examples in the documentation for this display platform. If by "previous functionality" you mean the sensor part, yes, that is not affected at all as long as the automatic disconnection is not disabled.

File: esphome/components/pvvx_mithermometer/display/__init__.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+async def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    await cg.register_component(var, config)
+    await display.register_display(var, config)
+    await ble_client.register_ble_node(var, config)
+    cg.add(var.set_disconnect_delay(config[CONF_DISCONNECT_DELAY].total_milliseconds))
+
+    if CONF_AUTO_CLEAR_ENABLED in config:
+        cg.add(var.set_auto_clear(config[CONF_AUTO_CLEAR_ENABLED]))
```
Comment (jesserockz) [inline esphome/components/pvvx_mithermometer/display/__init__.py]: A default is provided so the config options is always there
```suggestion
    cg.add(var.set_auto_clear(config[CONF_AUTO_CLEAR_ENABLED]))
```

File: esphome/components/pvvx_mithermometer/display/__init__.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    if CONF_TIME_ID in config:
+        time_ = await cg.get_variable(config[CONF_TIME_ID])
+        cg.add(var.set_time(time_))
+
+    if CONF_VALIDITY_PERIOD in config:
+        cg.add(var.set_validity_period(config[CONF_VALIDITY_PERIOD].total_seconds))
```
Comment (jesserockz) [inline esphome/components/pvvx_mithermometer/display/__init__.py]: Same here
```suggestion
    cg.add(var.set_validity_period(config[CONF_VALIDITY_PERIOD].total_seconds))
```

**Review by jesserockz — CHANGES_REQUESTED** [review]
Just a couple of minor changes, otherwise looks good.

**Comment by puuu** [comment]
@jesserockz  Sorry for my late reply.  The last commit introduces your minor changes.

**Comment by seeschloss** [comment]
Great, I was working on the same kind of thing but using the 0x60 message (to choose each segment instead of sending higher-level data, because I wanted to display a few letters as well).

But I don't have time to make it clean enough to release it, so I'm glad you did this and I hope it gets merged soon. When I have time (not before a few months though) I might add a way to use the 0x60 to this component.

Anyway thanks for your work!

**Comment by puuu** [comment]
@seeschloss thank you for your comment. I was not aware about the 0x60 message, but isn't it overwritten on every display refresh?

@jesserockz  what is the problem with this pull request? Why isn't it merged? I've made all the requested changes and the documentation has also been improved a long time ago.

**Comment by nagyrobi** [comment]
@puuu he's alone doing it

**Comment by seeschloss** [comment]
@puuu the 0x60 message stays displayed until you send a 0x61. A problem with it though that I just thought about, is that segments are different between devices so it would be complicated to use in a generic way (I didn't really have the problem since I only have one kind of device).

**Comment by puuu** [comment]
Unfortunately, this pull request is IMHO ready to merge, but nothing happens. The bot was already trying to close the documentation pull request again, so I will pump this pull request as well.

**Review by jesserockz — APPROVED** [review]

**Pull request merged**